Azure OIDC configuration instructions (#2282)
kubeflow/website: content/en/docs/azure/authentication-oidc.md

+++

> title = "Authentication using OIDC"

Reviewer: Suggestion:

- To help readers find this in search engines:

```suggestion
title = "Authentication using OIDC in Azure"
```

description = "Authentication and authorization support through OIDC for Kubeflow in Azure"
weight = 6
+++

> This section shows the how to setup Kubeflow with authentication and authorization support through OIDC in Azure using [Azure Active Directory](https://azure.microsoft.com/en-us/services/active-directory/).

Author: I'll keep setup to be consistent

## Prerequisites

- Get all prerequisite described in [Install Kubeflow](/docs/azure/deploy/install-kubeflow)
- [Register an application with the Microsoft Identity Platform](https://docs.microsoft.com/en-us/azure/active-directory/develop/quickstart-register-app#register-an-application)
- [Add a client secret](https://docs.microsoft.com/en-us/azure/active-directory/develop/quickstart-register-app#add-a-client-secret)

  **Note:**  Save client id, client secret, and tenant id in a secure place to be used in the next steps to configure OIDC Auth Service.

## Kubeflow configuration

> 1. Download the kfctl {{% kf-latest-version %}} release from the [Kubeflow releases page](<https://github.com/kubeflow/kfctl/releases/tag/>{{% kf-latest-version %}}).

Author: done

    - oidc_provider=https://login.microsoftonline.com/<tenant_id>/v2.0
    - oidc_redirect_uri=https://<load_balancer_ip> or dns_name>/login/oidc
    - oidc_auth_url=https://login.microsoftonline.com/<tenant_id>/oauth2/v2.0/authorize
    - application_secret=<client_secret>
    - skip_auth_uri=
    - namespace=istio-system
    - userid-header=kubeflow-userid
    - userid-prefix=
    ```

1. Configure OIDC scopes:

    In `/manifests/istio/oidc-authservice/base/statefulset.yaml` update [OIDC scopes](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-permissions-and-consent#openid-connect-scopes) to remove groups and keep profile and email.

    ```
    - name: OIDC_SCOPES
     value: "profile email"
    ```

3. Deploy Kubeflow:

    ```
    kfctl apply -V -f ${CONFIG_URI}
    ```

4. Check the resources deployed correctly in namespace `kubeflow`

    ```
    kubectl get all -n kubeflow
    ```

## Expose Kubeflow securely over HTTPS

1. Update Istio Gateway to expose port 443 with HTTPS and make port 80 redirect to 443:

    ```
    kubectl edit -n kubeflow gateways.networking.istio.io kubeflow-gateway
    ```

    The Gateway spec should look like the following:

>     {{< highlight yaml >}}

Reviewer 2: Is this highlight yams something website specific? It's not rendered good on GitHub

        name: http
        number: 80
        protocol: HTTP
              # Upgrade HTTP to HTTPS
        tls:
        httpsRedirect: true
    - hosts:
        - '*'
        port:
        name: https
        number: 443
        protocol: HTTPS
        tls:
        mode: SIMPLE
        privateKey: /etc/istio/ingressgateway-certs/tls.key
        serverCertificate: /etc/istio/ingressgateway-certs/tls.crt
    {{< /highlight >}}

1. Expose Kubeflow with a loadBalancer service:

    To expose Kubeflow with a LoadBalancer service, just change the type of the `istio-ingressgateway` service to `LoadBalancer`.

    ```
    kubectl patch service -n istio-system istio-ingressgateway -p '{"spec": {"type": "LoadBalancer"}}'
    ```

   After that, get the LoadBalancer's IP or Hostname from its status and create the necessary certificate.

    ```
    kubectl get svc -n istio-system istio-ingressgateway -o jsonpath='{.status.loadBalancer.ingress[0]}'
    ```

1. Create the Certificate with cert-manager:

    Create a new `certficate.yaml` with below yaml to create a self-signed Certificate with the cert-manager. For production environment please use appropriate trusted CA Certificate.

>     {{< highlight yaml >}}

Reviewer 2: Is this highlight yams something website specific? It's not rendered good on GitHub

    apiVersion: cert-manager.io/v1alpha2
    kind: Certificate
    metadata:
    name: istio-ingressgateway-certs
    namespace: istio-system
    spec:
    commonName: istio-ingressgateway.istio-system.svc
        # Use ipAddresses if your LoadBalancer issues an IP
    ipAddresses:
    - <LoadBalancer IP>
        # Use dnsNames if your LoadBalancer issues a hostname
    dnsNames:
    - <LoadBalancer HostName>
    isCA: true
    issuerRef:
        kind: ClusterIssuer
        name: kubeflow-self-signing-issuer
    secretName: istio-ingressgateway-certs
    {{< /highlight >}}

    Apply `certificate.yaml` in `istio-system` namespace

    ```
    kubectl apply -f certificate.yaml -n istio-system
    ```

>     After applying the above Certificate, cert-manager will generate the TLS certificate inside the istio-ingressgateway-certs secrets. The istio-ingressgateway-certs secret is mounted on the istio-ingressgateway deployment and used to serve HTTPS.

Author: done

>     After applying the above Certificate, cert-manager will generate the TLS certificate inside the istio-ingressgateway-certs secrets. The istio-ingressgateway-certs secret is mounted on the istio-ingressgateway deployment and used to serve HTTPS.

Reviewer: One more thing here, since you're setting `secretName` to `istio-ingressgateway-certs` in that YAML, let's use backticks (it helps new DevOps users visually):

```suggestion
    After applying the above Certificate, cert-manager will generate the TLS certificate inside the `istio-ingressgateway-certs` secrets. The `istio-ingressgateway-certs` secret is mounted on the `istio-ingressgateway` deployment and used to serve HTTPS.
```

1. [Configure Redirect URI for App](https://docs.microsoft.com/en-us/azure/active-directory/develop/quickstart-register-app#add-a-redirect-uri)

   Add below redirect uri to the app registered with Microsoft Identity:

   `https://<LoadBalancer IP or DNS Name>/login/oidc`

>    **Note:** Make sure the redirect uri for the app matches the `oidc_redirect_uri` value in OIDC auth service settings.

Author: done

   Navigate to `https://<LoadBalancer IP or DNS Name>/` and start using Kubeflow.

## Autheniticate Kubeflow pipelines using [Kubeflow Pipelines SDK](https://www.kubeflow.org/docs/pipelines/sdk/sdk-overview/)

    {{< highlight python >}}

>     import kfp

Reviewer 2: Is this highlight yams something website specific? It's not rendered good on GitHub

>     authservice_session_cookie='authservice_session=<cookie>'

Reviewer 2: Perhaps we need to clarify how to get this cookie?

>     client = kfp.Client(host='https://<LoadBalancer Address or DNS Name>/pipeline', cookies=authservice_session_cookie)

Author: done

    client.list_experiments(namespace="<your_namespace>")
    {{< /highlight >}}

>    **Limitation:** The current OIDC auth service in Kubeflow system supports only Authorization Code Flow. We are working on enabling more auth flows to support id-token and access-token based authentication and authorization.

Reviewer: Suggestion:

- Change "we" -> "the Azure team"
- Add a link to Auth Code Flow for those who aren't familiar with it
- Add a link to id_token and access_token and use underscore instead of dash

```suggestion
   **Limitation:** The current OIDC auth service in Kubeflow system supports only Authorization Code Flow. We are working on enabling more auth flows to support id-token and access-token based authentication and authorization.
    **Limitation:** The current OIDC auth service in Kubeflow system only
   supports [Authorization Code
   Flow](https://auth0.com/docs/flows/authorization-code-flow). The Azure team
   is working on enabling more auth flows to support [id_token and
   access_token](https://docs.microsoft.com/en-us/azure/active-directory/develop/access-tokens#validating-tokens)
   based authentication and authorization.
```

+++
title = "Authentication using OIDC in Azure"
description = "Authentication and authorization support through OIDC for Kubeflow in Azure"
weight = 6
+++

> This section shows the how to setup Kubeflow with authentication and authorization support through OIDC in Azure using [Azure Active Directory](https://azure.microsoft.com/en-us/services/active-directory/).

Reviewer: Let's still use "to set up" ([source](https://www.merriam-webster.com/dictionary/setup)) here, you may have missed it:

```suggestion
This section shows the how to set up Kubeflow with authentication and authorization support through OIDC in Azure using [Azure Active Directory](https://azure.microsoft.com/en-us/services/active-directory/).
```

## Prerequisites

- Install the [prerequisites for Kubeflow in Azure](/docs/azure/deploy/install-kubeflow)
- [Register an application with the Microsoft Identity Platform](https://docs.microsoft.com/en-us/azure/active-directory/develop/quickstart-register-app#register-an-application)
- [Add a client secret](https://docs.microsoft.com/en-us/azure/active-directory/develop/quickstart-register-app#add-a-client-secret)

>   **Note:**  Save your client Id, client secret, and tenant Id in a secure place to be used in the next steps to configure OIDC Auth Service.

Reviewer: I think this should be "ID", as mentioned in the previous review:

```suggestion
  **Note:**  Save your client ID, client secret, and tenant ID in a secure place to be used in the next steps to configure OIDC Auth Service (source: [search results[(https://docs.microsoft.com/en-us/search/?terms=ID&scope=Style%20guide) from Microsoft).
```

## Kubeflow configuration

1. Download the latest kfctl {{% kf-latest-version %}} release from the [Kubeflow releases page](<https://github.com/kubeflow/kfctl/releases/tag/>{{% kf-latest-version %}}).

1. Unpack the tar ball:

    ```
    tar -xvf kfctl_{{% kf-latest-version %}}_<platform>.tar.gz
    ```

> 1. Run the below commands to build configuration files before deploying Kubeflow. The code below includes an optional command to add the binary kfctl to your path. If you don’t add the binary to your path, you must use the full path to the kfctl binary each time you run it.

Reviewer: There's still repetition here which stands out when reading the instructions: 

`...to add the binary kfctl to your path. If you don’t add the binary to your path,...`.

Maybe, as mentioned before, use:

`...to add the binary kfctl to your path - if you don’t do it,...`

WDYT?

    ```
    kfctl apply -V -f ${CONFIG_URI}
    ```

4. Check that the resources were deployed correctly in namespace `kubeflow`:

    ```
    kubectl get all -n kubeflow
    ```

## Expose Kubeflow securely over HTTPS

1. Update Istio Gateway to expose port 443 with HTTPS and make port 80 redirect to 443:

    ```
    kubectl edit -n kubeflow gateways.networking.istio.io kubeflow-gateway
    ```

    The Gateway spec should look like the following:

    {{< highlight yaml >}}
    apiVersion: networking.istio.io/v1alpha3
    kind: Gateway
    metadata:
    name: kubeflow-gateway
    namespace: kubeflow
    spec:
    selector:
        istio: ingressgateway
    servers:
    - hosts:
        - '*'
        port:
        name: http
        number: 80
        protocol: HTTP
              # Upgrade HTTP to HTTPS
        tls:
        httpsRedirect: true
    - hosts:
        - '*'
        port:
        name: https
        number: 443
        protocol: HTTPS
        tls:
        mode: SIMPLE
        privateKey: /etc/istio/ingressgateway-certs/tls.key
        serverCertificate: /etc/istio/ingressgateway-certs/tls.crt
    {{< /highlight >}}

1. Expose Kubeflow with a load balancer service:

>     To expose Kubeflow with a LoadBalancer service, change the type of the `istio-ingressgateway` service to `LoadBalancer`.

Reviewer: ```suggestion
    To expose Kubeflow with a load balancer service, change the type of the `istio-ingressgateway` service to `LoadBalancer`.
```

(As per the Microsoft [docs](https://docs.microsoft.com/en-us/azure/architecture/guide/technology-choices/load-balancing-overview))

    ```
    kubectl patch service -n istio-system istio-ingressgateway -p '{"spec": {"type": "LoadBalancer"}}'
    ```

>    After that, obtain the LoadBalancer's IP or Hostname from its status and create the necessary certificate.

Reviewer: As per the Microsoft [docs](https://docs.microsoft.com/en-us/azure/architecture/guide/technology-choices/load-balancing-overview), since you're using the service name, let's add backticks (code font) here:

```suggestion
   After that, obtain the `LoadBalancer` IP or Hostname from its status and create the necessary certificate.
```
